common/flatpak-utils.c

```diff
@@ -5791,160 +5791,246 @@ get_write_progress (guint outstanding_writes)
 }
 
 static void
+get_average_bytes_per_content_object (OstreeAsyncProgress *progress,
+                                      guint64             *fetched_content_bytes,
+                                      guint64             *total_content_bytes)
+{
+  gint64 bytes_transferred, metadata_bytes, content_bytes;
+  gint fetched_metadata, outstanding_metadata, total_metadata;
+  gint total_deltas, fetched_deltas, fetched_deltas_size;
+  gint fetched_content = 0, total_content = 0;
+  gint fetched, total_fetches;
+
+  /* It would've been so much easier of OSTree just exposed all
+   * the progress report fields. All the following math is just
+   * the preparation for us to be able to do:
+   *
+   *   average_bytes_per_object = downloaded content bytes / downloaded content objects
+   */
+
+  /* All objects */
+  fetched = ostree_async_progress_get_uint (progress, "fetched");
+  total_fetches = fetched + ostree_async_progress_get_uint (progress, "outstanding-fetches");
+
+  /* Metadata objects */
+  fetched_metadata = ostree_async_progress_get_uint (progress, "metadata-fetched");
+  outstanding_metadata = ostree_async_progress_get_uint (progress, "outstanding-metadata-fetches");
+  total_metadata = fetched_metadata + outstanding_metadata;
+
+  /* Static deltas */
+  total_deltas = ostree_async_progress_get_uint (progress, "total-delta-parts")
+               + ostree_async_progress_get_uint (progress, "total-delta-fallbacks");
+  fetched_deltas = ostree_async_progress_get_uint (progress, "fetched-delta-parts")
+                 + ostree_async_progress_get_uint (progress, "fetched-delta-fallbacks");
+  fetched_deltas_size = ostree_async_progress_get_uint64 (progress, "fetched-delta-part-size");
+
+  /* All the downloaded bytes, without the downloaded deltas */
+  metadata_bytes = ostree_async_progress_get_uint64 (progress, "last-metadata-bytes");
+  bytes_transferred = ostree_async_progress_get_uint64 (progress, "bytes-transferred");
+  bytes_transferred -= fetched_deltas_size;
+
+  /* While downloading metadata, store the currently downloaded size
+   * and don't report anything */
+  if (outstanding_metadata > 0 || metadata_bytes == 0)
+    {
+      /* Keep track of when OSTree started to download content objects */
+      ostree_async_progress_set_uint64 (progress, "last-metadata-bytes", bytes_transferred);
+      return;
+    }
+
+  /* Finally, the content-related stuff */
+  content_bytes = bytes_transferred - metadata_bytes;
+  total_content = total_fetches - total_metadata - total_deltas;
```

Comment:
Should you really remove total_deltas here? Delta parts don't seem to be counted in total_fetches.